### PR TITLE
don't fail backfill due to DagsterAssetBackfillDataLoadError

### PR DESCRIPTION
**Pull request by johannkm**
https://elementl-workspace.slack.com/archives/C05MDAH8QM6/p1692105811624019

### CODE REVIEW

**Comment by johannkm** [comment]
Current dependencies on/for this PR:
* master
  * **PR #15860** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15860" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/15860?utm_source=stack-comment).

File: python_modules/dagster/dagster/_daemon/backfill.py [hunk 1]
```diff
@@ -38,10 +40,14 @@ def execute_backfill_iteration(
...
                 )
-        except Exception:
+        except Exception as e:
             error_info = serializable_error_info_from_exc_info(sys.exc_info())
-            instance.update_backfill(
-                backfill.with_status(BulkActionStatus.FAILED).with_error(error_info)
-            )
-            logger.error(f"Backfill failed for {backfill.backfill_id}: {error_info.to_string()}")
+            if os.environ.get("RETRY_CODE_LOCATION_ERROR") and isinstance(e, DagsterAssetBackfillDataLoadError):
+                logger.error(f"Backfill error for {backfill.backfill_id}: {error_info.to_string()}. Will retry.")
```
Comment (johannkm) [inline python_modules/dagster/dagster/_daemon/backfill.py:46]: I don't think we need to now, but in the future we could update the error here and make the ui surface an error even while the backfill isn't failed

File: python_modules/dagster/dagster/_daemon/backfill.py [hunk 2]
```diff
@@ -38,10 +40,14 @@ def execute_backfill_iteration(
...
                     backfill, logger, workspace_process_context, debug_crash_flags, instance
                 )
-        except Exception:
+        except Exception as e:
             error_info = serializable_error_info_from_exc_info(sys.exc_info())
-            instance.update_backfill(
-                backfill.with_status(BulkActionStatus.FAILED).with_error(error_info)
-            )
-            logger.error(f"Backfill failed for {backfill.backfill_id}: {error_info.to_string()}")
+            if os.environ.get("RETRY_CODE_LOCATION_ERROR") and isinstance(e, DagsterAssetBackfillDataLoadError):
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_daemon/backfill.py:45]: the name here seems very general for a fairly specific error

**Review by gibsondan — CHANGES_REQUESTED** [review]
do we need to raise and catch the exception here? Or could we modify the place where we were raising the exception? https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/execution/asset_backfill.py#L523-L530

Could that callsite log a warning and continue on instead of raising something that gets bubbled up all the way?

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py [hunk 3]
```diff
@@ -627,7 +628,15 @@ def execute_asset_backfill_iteration(
             if unloadable_locations
             else ""
         )
-        raise DagsterAssetBackfillDataLoadError(f"{ex}. {unloadable_locations_error}")
+        if os.environ.get("DAGSTER_BACKFILL_RETRY_DEFINITION_CHANGED_ERROR"):
+            logger.error(
+                "Retrying error in backfill"
+                f" {backfill.backfill_id}:\n{ex}. {unloadable_locations_error}"
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/execution/asset_backfill.py:634]: Language tweak suggestion to mimic the corresponding feature in the scheduler that retries once user code is available again:


"Backfill was unable to continue due to a missing asset or partition in the asset graph. The backfill will resume once it is available again. {unloadable_locations_error}"

**Review by gibsondan — APPROVED** [review]

**Pull request merged**